File: Kotlin/src/main/kotlin/beginner/h_DivideAndConquer/boj_1992.kt
package beginner.h_DivideAndConquer


private fun solve(arr: Array<IntArray>, row: Int, col: Int, size: Int) {
    fun check_arr(): Boolean {
        for (i in row until row + size) {
            val tmp = arr[row][col]

            for (j in col until col + size)
                if (arr[i][j] != tmp)
                    return false
        }

        return true
    }

    if (check_arr()) {
        print("${arr[row][col]}")
        return
    }

    val tmp = size / 2;

    print("(")
    solve(arr, row, col, tmp)
    solve(arr, row, col + tmp, tmp)
    solve(arr, row + tmp, col, tmp)
    solve(arr, row + tmp, col + tmp, tmp)
    print(")")
}

fun boj_1992() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()

    val arr = Array(n){ intArrayOf() }
        .map { readLine().map { it.digitToInt() }.toIntArray() }
        .toTypedArray()

    solve(arr, 0, 0, n)
}

File: v1/Cpp/boj/Beginner/08_DivideAndConquer/boj_1992.cpp
//
// Created by seungsu on 11/28/22.
//

#include <iostream>
#include <vector>

using namespace std;

static bool check_arr(vector<vector<int>>& v, int row, int col, int size) {
    for (int i = row; i < row + size; i++) {
        int val = v[row][col];

        for (int j = col; j < col + size; j++)
            if (v[i][j] != val)
                return false;
    }

    return true;
}

static void solve(vector<vector<int>>& v, int row, int col, int size) {
    if (check_arr(v, row, col, size)) {
        cout << v[row][col];
        return;
    }

    size /= 2;

    cout << "(";
    solve(v, row, col, size); // 왼쪽 위
    solve(v, row, col + size, size); // 오른쪽 위
    solve(v, row + size, col, size);      // 왼쪽 아래
    solve(v, row + size, col + size, size);    // 오른쪽 아래
    cout << ")";
}

void boj_1992() {
    int n;
    cin >> n;

    vector<vector<int>> v(n, vector<int>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            scanf("%1d", &v[i][j]);

    solve(v, 0, 0, n);
}